src/object/srv_obj.c

```diff
@@ -1638,6 +1643,187 @@ orf_to_dtx_epoch_flags(enum obj_rpc_flags orf_flags)
+}
+
+void
+ds_obj_ec_agg_handler(crt_rpc_t *rpc)
+{
+	struct obj_ec_agg_in	*oea = crt_req_get(rpc);
+	struct obj_ec_agg_out	*oeao = crt_reply_get(rpc);
+	daos_key_t		*dkey;
+	struct daos_oclass_attr	*oca;
+	struct bio_desc		*biod;
+	daos_iod_t		 iod = { 0 };
+	daos_recx_t		 recx = { 0 };
+	struct obj_io_context	 ioc;
+	daos_handle_t		 ioh = DAOS_HDL_INVAL;
+	int			 rc;
+
+	D_ASSERT(oea != NULL);
+	D_ASSERT(oeao != NULL);
+	D_ASSERT(daos_oclass_is_ec(oea->ea_oid.id_pub, &oca));
+
+	rc = obj_ioc_begin(oea->ea_map_ver, oea->ea_pool_uuid, oea->ea_coh_uuid,
+			   oea->ea_cont_uuid, opc_get(rpc->cr_opc), &ioc);
+
+	if (rc)	{
+		D_ERROR("ioc_begin failed: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+	dkey = (daos_key_t *)&oea->ea_dkey;
+	iod.iod_name = oea->ea_akey;
+	iod.iod_type = DAOS_IOD_ARRAY;
+	iod.iod_size = oea->ea_rsize;
+	iod.iod_nr = 1;
+	if (iod.iod_size) {
+		recx.rx_idx = (oea->ea_stripenum * oca->u.ec.e_len) |
+			PARITY_INDICATOR;
+		recx.rx_nr = oca->u.ec.e_len;
+		iod.iod_recxs = &recx;
+		rc = vos_update_begin(ioc.ioc_coc->sc_hdl, oea->ea_oid,
+				      oea->ea_epoch_range.epr_hi, 0, dkey, 1,
+				      &iod, NULL, NULL, 0, &ioh, NULL);
+		if (rc) {
+			D_ERROR(DF_UOID" Update begin failed: "DF_RC"\n",
+				DP_UOID(oea->ea_oid), DP_RC(rc));
+			goto out;
+		}
+		biod = vos_ioh2desc(ioh);
+		rc = bio_iod_prep(biod);
+		if (rc) {
+			D_ERROR(DF_UOID" bio_iod_prep failed: "DF_RC".\n",
+				DP_UOID(oea->ea_oid), DP_RC(rc));
+			goto out;
+		}
+		rc = obj_bulk_transfer(rpc, CRT_BULK_GET, false, &oea->ea_bulk,
+				       NULL, ioh, NULL, NULL, 1, NULL);
+		if (rc) {
+			D_ERROR(DF_UOID" bulk transfer failed: "DF_RC".\n",
+				DP_UOID(oea->ea_oid), DP_RC(rc));
+			goto out;
+		}
+
+		rc = bio_iod_post(biod);
+		if (rc) {
+			D_ERROR(DF_UOID" bio_iod_post failed: "DF_RC".\n",
+				DP_UOID(oea->ea_oid), DP_RC(rc));
+			goto out;
+		}
+		rc = vos_update_end(ioh, ioc.ioc_map_ver, dkey, rc, NULL);
+		if (rc) {
+			D_ERROR(DF_UOID" vos_update_end failed: "DF_RC".\n",
+				DP_UOID(oea->ea_oid), DP_RC(rc));
+			goto out;
+		}
+	}
+	if (oea->ea_remove_nr) {
+		daos_epoch_range_t	epr;
+		int			i;
+
+		for (i = 0; i < oea->ea_remove_nr; i++) {
+			epr.epr_hi = epr.epr_lo =
+				oea->ea_remove_eps.ca_arrays[i];
+			rc = vos_obj_array_remove(ioc.ioc_coc->sc_hdl,
+						  oea->ea_oid, &epr, dkey,
+						  &oea->ea_akey,
+						  &oea->
+						  ea_remove_recxs.ca_arrays[i]);
+			if (rc) {
+				D_ERROR(DF_UOID"array_remove failed: "DF_RC"\n",
+					DP_UOID(oea->ea_oid), DP_RC(rc));
+			}
+		}
+
+	} else {
+		recx.rx_idx = oea->ea_stripenum * oca->u.ec.e_len *
+			oca->u.ec.e_k;
+		recx.rx_nr = oca->u.ec.e_k * oca->u.ec.e_len;
+		rc = vos_obj_array_remove(ioc.ioc_coc->sc_hdl, oea->ea_oid,
+					  &oea->ea_epoch_range, dkey,
+					  &oea->ea_akey, &recx);
+		if (rc) {
+			D_ERROR(DF_UOID"array_remove failed: "DF_RC"\n",
+				DP_UOID(oea->ea_oid), DP_RC(rc));
+		}
+	}
+out:
+	obj_rw_reply(rpc, rc, 0, &ioc);
+	obj_ioc_end(&ioc, rc);
+}
```

Comment:
probably put these two functions to srv_ec_aggregate.c as well?

Reply:
I certainly can.  I thought it was convention to put all server handlers for objects in srv_obj.c, but I'm happy to move them.